### PR TITLE
Implement PhysicsParticleDescriptor and PhysicsParticleRegister

### PR DESCRIPTION
**Pull request by chongchonghe**
### Description
This PR implements `PhysicsParticleDescriptor` and `PhysicsParticleRegister` (in PhysicsParticles.hpp) to register any particle types into `particleRegistry_`. Then, you can loop over all particles in `particleRegister_` to do actions (e.g. deposit gravity) on all particle types with a certain flag. 

#### Particle definition

Line 2068 of simulation.hpp:

```cpp
template <typename problem_t> void AMRSimulation<problem_t>::InitPhyParticles()
{
	if (do_rad_particles != 0) {
		AMREX_ASSERT(RadParticles == nullptr);

		// Create radiating particle descriptor
		// the indices are: massIndex_, lumIndex_, interactsWithHydro_
		auto radParticleDesc = std::make_unique<quokka::PhysicsParticleDescriptor>(-1, quokka::RadParticleLumIdx, false);

		// Create particle container
		RadParticles = std::make_unique<quokka::RadParticleContainer<problem_t>>(this);
		RadParticles->SetVerbose(0);

		// Set container in descriptor
		radParticleDesc->neighborParticleContainer_ = RadParticles.get();

		// Register with particle register
		particleRegister_->registerParticleType("Rad_particles", std::move(radParticleDesc));

		// Initialize particles through derived class
		createInitialRadParticles();

		RadParticles->Redistribute();
	}

	if (do_cic_particles != 0) {
    // repeating the above for CICParticles
    ...
  }
}

```

#### Shared functions in PhysicsParticleRegister

`PhysicsParticleRegister` has the following shared functions that are used among all/some particle types: `depositRadiation, depositMass, redistribute, writePlotFile, writeCheckpoint`. 

Example 1: The following code deposit mass/gravity of all particles that belong to a particle type (e.g. 'CIC_particles') with `mass_idx >= 0`. This operates on all AMR levels.

```cpp
particleRegister_->depositMass(rhs, finest_level, Gconst_);
```

source code:

```cpp
void depositMass(amrex::Vector<amrex::MultiFab>& rhs, int finest_lev, amrex::Real Gconst)
{
for (const auto &[name, descriptor] : particleRegistry_) {
	if (descriptor->getMassIndex() >= 0) {
		// Get particle container and deposit mass
		auto *container = static_cast<RadParticleContainer<problem_t> *>(descriptor->getParticleContainer());
		if (container != nullptr) {
			amrex::ParticleToMesh(*container, amrex::GetVecOfPtrs(rhs), 0, finest_lev, 
				MassDeposition{Gconst, descriptor->getMassIndex(), 0, 1}, true);
		}
	}
}
}
```

Example 2: The following code write checkpoint files of all particles of all types, each in their own folder. 

```cpp
// write all particles in particleRegister_ to checkpoint file
particleRegister_->writeCheckpoint(checkpointname, true);
```

#### Test problems

`RadParticle` and `BinaryOrbitCIC`. Passing both tests. 

### Related issues

Are there any GitHub issues that are fixed by this pull request? Add a link to them here.

### Checklist
_Before this pull request can be reviewed, all of these tasks should be completed. Denote completed tasks with an `x` inside the square brackets `[ ]` in the Markdown source below:_
- [x] I have added a description (see above).
- [x] I have added a link to any related issues (if applicable; see above).
- [x] I have read the [Contributing Guide](https://github.com/quokka-astro/quokka/blob/development/CONTRIBUTING.md).
- [x] I have added tests for any new physics that this PR adds to the code.
- [ ] *(For quokka-astro org members)* I have manually triggered the GPU tests with the magic comment `/azp run`.

### CODE REVIEW

**Comment by chongchonghe** [comment]
Next step:
- Test on defining a RadCICParticles type that have both luminosity and mass. 
- Update `kickParticlesAllLevels` and `driftParticlesAllLevels` to use the new particle architecture. 

A few things that can be improved (I don't have ideas yet):
- Function `InitPhyParticles()` has lots of repetitions. Ideally, the particle initialization process can be simplified as `initializeParticleType("CIC_particles", CICParticles, 0, -1, false)` and `initializeParticleType("Rad_particles", RadParticles, -1, quokka::RadParticleLumIdx, false)`. 
- Reading particle data from chkfiles can be simplified similarly. 

@BenWibking @markkrumholz Please take a look at this PR and let me know what you think.

File: src/simulation.hpp [hunk 1]
```diff
@@ -2054,14 +2065,49 @@ template <typename problem_t> void AMRSimulation<problem_t>::InitParticles()
...
+		// Create radiating particle descriptor
+		// the indices are: mass, lum, hydro_interact
+		auto radParticleDesc = std::make_unique<quokka::PhysicsParticleDescriptor>(-1, quokka::RadParticleLumIdx, false);
+
+		// Create particle container
+		RadParticles = std::make_unique<quokka::RadParticleContainer<problem_t>>(this);
+		RadParticles->SetVerbose(0);
+
+		// Set container in descriptor
+		radParticleDesc->neighborParticleContainer_ = RadParticles.get();
```
Comment (BenWibking) [inline src/simulation.hpp:2086]: This does not transfer ownership of `RadParticles`. In your implementation, `RadParticles` will be deallocated on line 2091, and attempting to access `radParticleDesc->neighborParticleContainer_` will cause a segmentation fault.

This is why I am so insistent about using `std::unique_ptr` instead of raw pointers. You cannot have this bug if `radParticleDesc->neighborParticleContainer` is a `std::unique_ptr`.

File: src/particles/PhysicsParticles.hpp [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+	int lumIndex_{-1};		 // index for radiation luminosity, -1 if not used
+	bool interactsWithHydro_{false}; // whether particles interact with hydro
+
+      public:
+	PhysicsParticleDescriptor(int mass_idx, int lum_idx, bool hydro_interact)
+	    : massIndex_(mass_idx), lumIndex_(lum_idx), interactsWithHydro_(hydro_interact)
+	{
+	}
+	virtual ~PhysicsParticleDescriptor() = default;
+	void *neighborParticleContainer_{}; // pointer to particle container, type-erased
```
Comment (BenWibking) [inline src/particles/PhysicsParticles.hpp]: There is no reason to do type-erasure here, and no reason to make this a raw pointer.

File: src/particles/PhysicsParticles.hpp [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+	// Getters
+	[[nodiscard]] auto getMassIndex() const -> int { return massIndex_; }
+	[[nodiscard]] auto getLumIndex() const -> int { return lumIndex_; }
+	[[nodiscard]] auto getInteractsWithHydro() const -> bool { return interactsWithHydro_; }
+
+	// Virtual methods that can be overridden
+	virtual void hydroInteract() {} // Default no-op
```
Comment (BenWibking) [inline src/particles/PhysicsParticles.hpp]: There is no reason to make this a virtual function. You can just use `std::function` instead.

File: src/particles/PhysicsParticles.hpp [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+	PhysicsParticleDescriptor &operator=(const PhysicsParticleDescriptor &) = delete;
+	PhysicsParticleDescriptor(PhysicsParticleDescriptor &&) = delete;
+	PhysicsParticleDescriptor &operator=(PhysicsParticleDescriptor &&) = delete;
+};
+
+// Registry for physics particles
+template <typename problem_t> class PhysicsParticleRegister
+{
+      private:
+	std::map<std::string, std::unique_ptr<PhysicsParticleDescriptor>> particleRegistry_;
```
Comment (BenWibking) [inline src/particles/PhysicsParticles.hpp:132]: There is no reason to make this a map of pointers. You can just make this `std::map<std::string, PhysicsParticleDescriptor>`.

**Comment by BenWibking** [comment]
Here is a good explanation of why void* is bad in C++: https://www.reddit.com/r/cpp_questions/comments/jaeky1/void_pointers/

File: src/particles/PhysicsParticles.hpp (hunk 2 above)
Comment (BenWibking) [reply src/particles/PhysicsParticles.hpp]: You can make this `std::unique_ptr<amrex::ParticleContainerBase>` and then `dynamic_cast` to the sub-types of particle container, as needed: https://amrex-codes.github.io/amrex/doxygen/classamrex_1_1ParticleContainerBase.html

File: src/particles/PhysicsParticles.hpp [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+		return nullptr;
+	}
+
+	// Deposit radiation from all particles that have luminosity
+	void depositRadiation(amrex::MultiFab &radEnergySource, int lev, amrex::Real current_time)
+	{
+		for (const auto &[name, descriptor] : particleRegistry_) {
+			if (descriptor->getLumIndex() >= 0) {
+				// Get particle container and deposit luminosity
+				auto *container = static_cast<RadParticleContainer<problem_t> *>(descriptor->neighborParticleContainer_);
```
Comment (BenWibking) [inline src/particles/PhysicsParticles.hpp]: Use `dynamic_cast` instead.

File: src/particles/PhysicsParticles.hpp [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+		}
+	}
+
+	// Deposit mass from all particles that have mass for gravity calculation
+	void depositMass(amrex::Vector<amrex::MultiFab> &rhs, int finest_lev, amrex::Real Gconst)
+	{
+		for (const auto &[name, descriptor] : particleRegistry_) {
+			if (descriptor->getMassIndex() >= 0) {
+				// Get particle container and deposit mass
+				auto *container = static_cast<RadParticleContainer<problem_t> *>(descriptor->neighborParticleContainer_);
```
Comment (BenWibking) [inline src/particles/PhysicsParticles.hpp]: Use `dynamic_cast` instead.

File: src/particles/PhysicsParticles.hpp [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+				}
+			}
+		}
+	}
+
+	// Run Redistribute(lev) on all particles in particleRegistry_
+	void redistribute(int lev)
+	{
+		for (const auto &[name, descriptor] : particleRegistry_) {
+			auto *container = static_cast<RadParticleContainer<problem_t> *>(descriptor->neighborParticleContainer_);
```
Comment (BenWibking) [inline src/particles/PhysicsParticles.hpp]: Use `dynamic_cast` instead.

File: src/particles/PhysicsParticles.hpp [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+				container->Redistribute(lev);
+			}
+		}
+	}
+
+	// Run Redistribute(lev, ngrow) on all particles in particleRegistry_
+	void redistribute(int lev, int ngrow)
+	{
+		for (const auto &[name, descriptor] : particleRegistry_) {
+			auto *container = static_cast<RadParticleContainer<problem_t> *>(descriptor->neighborParticleContainer_);
```
Comment (BenWibking) [inline src/particles/PhysicsParticles.hpp]: Use `dynamic_cast` instead.

File: src/particles/PhysicsParticles.hpp [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+				container->Redistribute(lev, container->finestLevel(), ngrow);
+			}
+		}
+	}
+
+	// Run WritePlotFile(plotfilename, name) on all particles in particleRegistry_
+	void writePlotFile(const std::string &plotfilename)
+	{
+		for (const auto &[name, descriptor] : particleRegistry_) {
+			auto *container = static_cast<RadParticleContainer<problem_t> *>(descriptor->neighborParticleContainer_);
```
Comment (BenWibking) [inline src/particles/PhysicsParticles.hpp]: Use `dynamic_cast` instead.

File: src/particles/PhysicsParticles.hpp [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+				container->WritePlotFile(plotfilename, name);
+			}
+		}
+	}
+
+	// Run Checkpoint(checkpointname, name, true) on all particles in particleRegistry_
+	void writeCheckpoint(const std::string &checkpointname, bool include_header)
+	{
+		for (const auto &[name, descriptor] : particleRegistry_) {
+			auto *container = static_cast<RadParticleContainer<problem_t> *>(descriptor->neighborParticleContainer_);
```
Comment (BenWibking) [inline src/particles/PhysicsParticles.hpp]: Use `dynamic_cast` instead.

File: src/particles/PhysicsParticles.hpp [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+	int massIndex_{-1};		 // index for gravity mass, -1 if not used
+	int lumIndex_{-1};		 // index for radiation luminosity, -1 if not used
+	bool interactsWithHydro_{false}; // whether particles interact with hydro
+
+      public:
+	PhysicsParticleDescriptor(int mass_idx, int lum_idx, bool hydro_interact)
+	    : massIndex_(mass_idx), lumIndex_(lum_idx), interactsWithHydro_(hydro_interact)
+	{
+	}
+	virtual ~PhysicsParticleDescriptor() = default;
```
Comment (BenWibking) [inline src/particles/PhysicsParticles.hpp]: No reason to make this virtual.

File: src/radiation/source_terms_multi_group.hpp [hunk 12]
```diff
@@ -616,9 +616,10 @@ void RadSystem<problem_t>::AddSourceTermsMultiGroup(array_t &consVar, arrayconst
 
 		// load radiation energy source term
 		// plus advection source term (for well-balanced/SDC integrators)
+		// Note that radEnergySource should contain the luminosity volume density, L / V; unit: erg s^-1 cm^-3
 		quokka::valarray<double, nGroups_> Src;
 		for (int g = 0; g < nGroups_; ++g) {
-			Src[g] = dt * (chat * radEnergySource(i, j, k, g));
+			Src[g] = dt * (chat / c * radEnergySource(i, j, k, g));
```
Comment (BenWibking) [inline src/radiation/source_terms_multi_group.hpp:622]: This seems like an unrelated change that should go in a separate PR.

File: src/simulation.hpp [hunk 13]
```diff
@@ -447,16 +449,22 @@ template <typename problem_t> class AMRSimulation : public amrex::AmrCore
 #ifdef AMREX_PARTICLES
 	void InitParticles();	 // create tracer particles
 	void InitCICParticles(); // create CIC particles
+	void InitPhyParticles(); // create PhysicsParticles
 	int do_tracers = 0;
 	int do_cic_particles = 0;
+	int do_rad_particles = 0;
 	std::unique_ptr<amrex::AmrTracerParticleContainer> TracerPC;
 	std::unique_ptr<quokka::CICParticleContainer> CICParticles;
+	std::unique_ptr<quokka::RadParticleContainer<problem_t>> RadParticles;
```
Comment (BenWibking) [inline src/simulation.hpp:458]: I think this doesn't need to be a pointer. It can just be `quokka::RadParticleContainer`.

File: src/simulation.hpp [hunk 14]
```diff
@@ -447,16 +449,22 @@ template <typename problem_t> class AMRSimulation : public amrex::AmrCore
...
 #endif
 
 	// external objects
 #ifdef AMREX_USE_ASCENT
 	Ascent ascent_;
 #endif
+
+	// Add PhysicsParticleRegister member
+	std::unique_ptr<quokka::PhysicsParticleRegister<problem_t>> particleRegister_;
```
Comment (BenWibking) [inline src/simulation.hpp:468]: This does not need to be a pointer. It can just be `quokka::PhysicsParticleRegister<problem_t>`.

File: src/simulation.hpp (hunk 13 above)
Comment (BenWibking) [inline src/simulation.hpp:458]: We don't need separate objects for these. They can just live inside the `particleRegister_`.

File: src/radiation/source_terms_single_group.hpp [hunk 15]
```diff
@@ -47,9 +47,12 @@ void RadSystem<problem_t>::AddSourceTermsSingleGroup(array_t &consVar, arraycons
 		const double Erad0 = consPrev(i, j, k, radEnergy_index);
 		AMREX_ASSERT(Erad0 > 0.0);
 
+		const double cscale = c / chat;
```
Comment (BenWibking) [inline src/radiation/source_terms_single_group.hpp:50]: The changes in this file look unrelated to particles. Can you put them in a separate PR?

File: src/particles/PhysicsParticles.hpp (hunk 4 above)
Comment (chongchonghe) [reply src/particles/PhysicsParticles.hpp:132]: particleRegistry_ has to be a map from string to std::unique_ptr because it's used to store a descriptor, which is a `std::unique_ptr<ParticleDescriptor>`

Comment (BenWibking) [reply src/particles/PhysicsParticles.hpp:132]: Ok, I don't follow. Why do the descriptors need to be `std::unique_ptr<ParticleDescriptor>` instead of just `ParticleDescriptor`?

File: src/radiation/source_terms_multi_group.hpp (hunk 12 above)
Comment (chongchonghe) [reply src/radiation/source_terms_multi_group.hpp:622]: This is related. I found the original definition of the source term (which I adapted from your original code) confusing. There is a factor of `c` difference from my current implementation. I assume radEnergySource contain the luminosity density (erg s^-1 cm^-3), so the conversion from radEnergySource to Src should be `Src = dt * (chat/c) * radEnergySource`.

File: src/radiation/source_terms_single_group.hpp (hunk 15 above)
Comment (chongchonghe) [reply src/radiation/source_terms_single_group.hpp:50]: This is used later to convert radEnergySource to Src, as I explained in my other comments.

File: src/simulation.hpp (hunk 13 above)
Comment (chongchonghe) [reply src/simulation.hpp:458]: This is the part I don't understand. I don't think defining RadParticles inside particleRegister_ will work, because that means RadParticles will have to be declared inside a constexpr if, which is not allowed. I haven't tried. 

I forgot to mention that I will change do_cic_particles, etc to constexpr's in Physics_Traits later.

**Comment by chongchonghe** [comment]
@BenWibking I tried to make `neighborParticleContainer_` a `std::unique_ptr<amrex::ParticleContainerBase>`, but it didn't work. The best I can do is to make it a non-owning pointer and it worked: `amrex::ParticleContainerBase *neighborParticleContainer_{nullptr};`. The reason `std::unique_prt` didn't work is because in `InitPhyParticles()`, if we attempt to do `radParticleDesc->neighborParticleContainer_ = std::unique_ptr<amrex::ParticleContainerBase>(RadParticles.get());`, it's trying to 'own' the memory, but the original `RadParticles` also owns the memory. This causes problems. With non-owning pointer and `radParticleDesc->neighborParticleContainer_ = RadParticles.get();`, this is no such a problem. So, in the current framework where `RadParticles` is defined in AMRSimulation (not in particleRegister_), we have to use non-owning pointer.

File: src/particles/PhysicsParticles.hpp (hunk 3 above)
Comment (chongchonghe) [reply src/particles/PhysicsParticles.hpp]: Done.

**Comment by BenWibking** [comment]
> @BenWibking I tried to make `neighborParticleContainer_` a `std::unique_ptr<amrex::ParticleContainerBase>`, but it didn't work. The best I can do is to make it a non-owning pointer and it worked: `amrex::ParticleContainerBase *neighborParticleContainer_{nullptr};`. The reason `std::unique_prt` didn't work is because in `InitPhyParticles()`, if we attempt to do `radParticleDesc->neighborParticleContainer_ = std::unique_ptr<amrex::ParticleContainerBase>(RadParticles.get());`, it's trying to 'own' the memory, but the original `RadParticles` also owns the memory. This causes problems. With non-owning pointer and `radParticleDesc->neighborParticleContainer_ = RadParticles.get();`, this is no such a problem. So, in the current framework where `RadParticles` is defined in AMRSimulation (not in particleRegister_), we have to use non-owning pointer.

Ok, that makes sense. I think I missed that in my first review of the PR, sorry. If it's defined/owned by AMRSimulation, then using a raw pointer inside the descriptor (as long as it's not a void pointer) is totally fine by me.

**Comment by chongchonghe** [comment]
There is a problem with my previous implementation. In the following `depositMass` function, I have to use `auto *container = dynamic_cast<RadParticleContainer<problem_t> *>(descriptor->neighborParticleContainer_);`, but this is not generic. There will be multiple particles types that will have mass (e.g. CICRadParticles), and the container won't be `RadParticleContainer`, but we have to know it in order to do dynamic_cast. 

```cpp
	// Deposit mass from all particles that have mass for gravity calculation
	void depositMass(amrex::Vector<amrex::MultiFab> &rhs, int finest_lev, amrex::Real Gconst)
	{
		for (const auto &[name, descriptor] : particleRegistry_) {
			if (descriptor->getMassIndex() >= 0) {
				// Get particle container and deposit mass
				auto *container = dynamic_cast<RadParticleContainer<problem_t> *>(descriptor->neighborParticleContainer_);
				if (container != nullptr) {
					amrex::ParticleToMesh(*container, amrex::GetVecOfPtrs(rhs), 0, finest_lev,
							      MassDeposition{Gconst, descriptor->getMassIndex(), 0, 1}, true);
				}
			}
		}
	}
```

We can do checks of dynamic_cast like in the examples here: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c146-use-dynamic_cast-where-class-hierarchy-navigation-is-unavoidable ,
```cpp
if (auto *container = dynamic_cast<CICParticleContainer<problem_t> *>(descriptor->neighborParticleContainer_)) {
	amrex::ParticleToMesh(*container, amrex::GetVecOfPtrs(rhs), 0, finest_lev, MassDeposition{Gconst, descriptor->getMassIndex(), 0, 1}, true);
} else if (auto *container = dynamic_cast<CICRadParticleContainer<problem_t> *>(descriptor->neighborParticleContainer_)) {
	amrex::ParticleToMesh(*container, amrex::GetVecOfPtrs(rhs), 0, finest_lev, MassDeposition{Gconst, descriptor->getMassIndex(), 0, 1}, true);
} else if (auto *container = dynamic_cast<SinkParticleContainer<problem_t> *>(descriptor->neighborParticleContainer_)) {
	amrex::ParticleToMesh(*container, amrex::GetVecOfPtrs(rhs), 0, finest_lev, MassDeposition{Gconst, descriptor->getMassIndex(), 0, 1}, true);
}
```

 but this is exactly the sort of thing we want to avoid by using this new particle architecture. 

-------------

My solution is to use type-erased pointer to the class operations, allowing it to work with any particle container type through a common interface. I created a new PR for this: #826 . Although it works, it adds a lot more complexity compared to this PR, so I don't know if I like it.

File: src/particles/PhysicsParticles.hpp (hunk 4 above)
Comment (chongchonghe) [reply src/particles/PhysicsParticles.hpp:132]: I believe changing it to `std::map<std::string, PhysicsParticleDescriptor>` would make a copy of PhysicsParticleDescriptor, which should be avoided. 

In simulation.hpp, radParticleDesc is created as a pointer, so it makes sense to define `particleRegistry_` as a map from string to a pointer. 
```cpp
auto radParticleDesc = std::make_unique<quokka::PhysicsParticleDescriptor>(-1, quokka::RadParticleLumIdx, quokka::RadParticleBirthTimeIdx, false);
// registration
particleRegister_->registerParticleType("Rad_particles", std::move(radParticleDesc));
```

Comment (BenWibking) [reply src/particles/PhysicsParticles.hpp:132]: You don't have to create it as a pointer. If you then `std::move` it, it won't be copied.

**Comment by BenWibking** [comment]
> There is a problem with my previous implementation. In the following `depositMass` function, I have to use `auto *container = dynamic_cast<RadParticleContainer<problem_t> *>(descriptor->neighborParticleContainer_);`, but this is not generic. There will be multiple particles types that will have mass (e.g. CICRadParticles), and the container won't be `RadParticleContainer`, but we have to know it in order to do dynamic_cast.
> 
> ```c++
> 	// Deposit mass from all particles that have mass for gravity calculation
> 	void depositMass(amrex::Vector<amrex::MultiFab> &rhs, int finest_lev, amrex::Real Gconst)
> 	{
> 		for (const auto &[name, descriptor] : particleRegistry_) {
> 			if (descriptor->getMassIndex() >= 0) {
> 				// Get particle container and deposit mass
> 				auto *container = dynamic_cast<RadParticleContainer<problem_t> *>(descriptor->neighborParticleContainer_);
> 				if (container != nullptr) {
> 					amrex::ParticleToMesh(*container, amrex::GetVecOfPtrs(rhs), 0, finest_lev,
> 							      MassDeposition{Gconst, descriptor->getMassIndex(), 0, 1}, true);
> 				}
> 			}
> 		}
> 	}
> ```
> 
> We can do checks of dynamic_cast like in the examples here: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c146-use-dynamic_cast-where-class-hierarchy-navigation-is-unavoidable ,
> 
> ```c++
> if (auto *container = dynamic_cast<CICParticleContainer<problem_t> *>(descriptor->neighborParticleContainer_)) {
> 	amrex::ParticleToMesh(*container, amrex::GetVecOfPtrs(rhs), 0, finest_lev, MassDeposition{Gconst, descriptor->getMassIndex(), 0, 1}, true);
> } else if (auto *container = dynamic_cast<CICRadParticleContainer<problem_t> *>(descriptor->neighborParticleContainer_)) {
> 	amrex::ParticleToMesh(*container, amrex::GetVecOfPtrs(rhs), 0, finest_lev, MassDeposition{Gconst, descriptor->getMassIndex(), 0, 1}, true);
> } else if (auto *container = dynamic_cast<SinkParticleContainer<problem_t> *>(descriptor->neighborParticleContainer_)) {
> 	amrex::ParticleToMesh(*container, amrex::GetVecOfPtrs(rhs), 0, finest_lev, MassDeposition{Gconst, descriptor->getMassIndex(), 0, 1}, true);
> }
> ```
> 
> but this is exactly the sort of thing we want to avoid by using this new particle architecture.
> 
> My solution is to use type-erased pointer to the class operations, allowing it to work with any particle container type through a common interface. I created a new PR for this: #826 . Although it works, it adds a lot more complexity compared to this PR, so I don't know if I like it.

No type-erasure is needed. I think there is a simple solution. See here: https://github.com/quokka-astro/quokka/pull/826#issuecomment-2551941829.

**Comment by chongchonghe** [comment]
This PR doesn't work. The other two PRs: #826 and #827 work, one using type-erased ParticleOperations and the other using derived ParticleDescriptors.

**Comment by BenWibking** [comment]
Ok. In that case, can you close this PR?
